### ISSUE TITLE
Add new resource to support firewall plus (NGFW Enterprise)

### ISSUE THREAD

**Issue by imrannayer** [hashicorp/terraform-provider-google#15779]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### New or Affected Resource(s)

* google_network_security_security_profile
* google_network_security_security_profile_group
* google_network_security_firewall_endpoint
* google_network_security_firewall_endpoint_association

### Potential Terraform Configuration

```tf
# Propose what you think the configuration to take advantage of this feature should look like.
# We may not use it verbatim, but it's helpful in understanding your intent.
```

### References

* [Security Profile](https://cloud.google.com/firewall/docs/configure-security-profiles)
* [Security profile group](https://cloud.google.com/firewall/docs/configure-security-profile-groups)
* [Firewall endpoint](https://cloud.google.com/firewall/docs/configure-firewall-endpoints)

API
* https://cloud.google.com/firewall/docs/reference/network-security/rest/v1beta1/organizations.locations.firewallEndpoints
* https://cloud.google.com/firewall/docs/reference/network-security/rest/v1beta1/projects.locations.firewallEndpointAssociations
* https://cloud.google.com/firewall/docs/reference/network-security/rest/v1beta1/organizations.locations.securityProfiles
* https://cloud.google.com/firewall/docs/reference/network-security/rest/v1beta1/organizations.locations.securityProfileGroups

b/303808491

**Comment by rileykarson** [comment]
Posting these because I got a little mixed up here- there's two `networksecurity.googleapis.com` API definitions. Weird! https://cloud.google.com/firewall/docs/reference/network-security/rest, https://cloud.google.com/secure-web-proxy/docs/reference/network-security/rest

**Comment by rileykarson** [comment]
Yep! What's unusual is that they share a service endpoint, I got mixed up looking for these resources because of that

**Comment by maxi-cit** [comment]
Hello folks, I will be working on these as part of the NGFW construct support

**Comment by LucaPrete** [comment]
FYI I'm going to send in the first PR for security profiles.

**Comment by maxi-cit** [comment]
Hey @LucaPrete, could you tell me which resource are you referring to, please? is it the google_network_security_security_profile? when are you opening this PR?

**Comment by LucaPrete** [comment]
Yes.
I’m verifying the latest tests now. Almost done

Il giorno mer 10 gen 2024 alle 17:40 Max Portocarrero CI&T <
***@***.***> ha scritto:

**Comment by maxi-cit** [comment]
Hey @LucaPrete, gotcha, its totally fine. the only issue is that I already had some effort on the same resource. So I recommend you to open a draft PR or claim for the specific resource before start working on the PR the way I did in previous comment. This way, the community does not duplicate the work.

PD: I will move to the firewall related resources and the other updates.

**Comment by paulomarquescit** [comment]
Hello
I am will start working on the security profile groups.

**Comment by LucaPrete** [comment]
Max, thanks!

Any chance we can first finalize the support to secure tags in gke node
pools?

Il giorno mer 10 gen 2024 alle 18:27 paulomarquescit <
***@***.***> ha scritto:

**Comment by maxi-cit** [comment]
Hello @LucaPrete Yes I’m verifying the latest tests. Almost done

I know I claimed for the firewall resources but did you have something on them already

PD: I am glad of your help on the product implementation. I just pointed out we need to sync our efforts

**Comment by LucaPrete** [comment]
Yep. Makes sense!

I thought you were not from Google!
Can you please contact me internally? I’m not able to find you.

-Luca

Il giorno mer 10 gen 2024 alle 19:08 Max Portocarrero CI&T <
***@***.***> ha scritto:

**Comment by LucaPrete** [comment]
> Hello I am will start working on the security profile groups. The google_network_security_security_profile_group

Hi @paulomarquescit I can do this today if for you is ok. Have you already started? Don't want to seem rude. I'm trying to speed up the process a bit... ;)

**Comment by paulomarquescit** [comment]
Hi @LucaPrete if you can finish today, you can continue work on the google_network_security_security_profile_group

**Comment by paulomarquescit** [comment]
Hello
I am will start working on the firewall endpoint resource.
The google_network_security_firewall_endpoint

**Comment by LucaPrete** [comment]
@paulomarquescit I agreed with @maxi-cit that I would have continued to work on this until Monday. Then I'll be out for 3 weeks :) I already have something in place for endpoints as well (PR coming). Sorry for the continuous overlaps. Just trying to commit whatever I can before my leave

**Comment by LucaPrete** [comment]
> @paulomarquescit I agreed with @maxi-cit that I would have continued to work on this until Monday. Then I'll be out for 3 weeks :) I already have something in place for endpoints as well (PR coming). Sorry for the continuous overlaps. Just trying to commit whatever I can before my leave

...Following up on this...it seems an interesting conversation to have. FirewallEndpoints currently take 60+ minutes to be created, so I'd exclude we want to run real tests for them. Also, product documentation recommends to create them asynchronously, given the time it takes to create them. It uses a special `requestId` urlParam for this, that users can optionally insert in each request. Anyway, I would think this is not applicable to terraform. What do you think @maxi-cit @paulomarquescit  @rileykarson ?

**Comment by maxi-cit** [comment]
Hello @LucaPrete , that was an issue Paulo told me about. Currently, I do not know of a way to customize the generated custom function other than increasing the timeout time or going for a hand written resource.

**Comment by LucaPrete** [comment]
So, I think I made it. Please, try to have a look.

I basically used the same approach we've been using for gcve clusters which
suffers from the same issue.
TL;DR I skip test in the examples and I use acctest.SkipIfVcr(t) in the
manual test.

The resource should be ready and I ran the manual test which passed.

-Luca

Il giorno ven 12 gen 2024 alle ore 16:35 Max Portocarrero CI&T <
***@***.***> ha scritto:

**Comment by melinath** [comment]
Reopening since not all resources are supported

**Comment by roaks3** [comment]
Sorry for the noise (I missed this before merging the most recent PR), reopening again since not all resources are supported

**Comment by LucaPrete** [comment]
@melinath actually we should reopen it, as we still miss endpoint associations (my pr under review) and a new action in firewall policies

**Comment by melinath** [comment]
whoops, thanks!

**Comment by imrannayer** [comment]
#17030 is needed for this to work

**Comment by melinath** [comment]
looks like we can track the remaining work in that issue.
